src/scancode/cli.py

```diff
@@ -217,724 +225,891 @@ def print_version(ctx, param, value):
+        if with_timing:
+            start = time()
         try:
-            scan_details = scanner(location)
-            # consume generators
-            if isinstance(scan_details, GeneratorType):
-                scan_details = list(scan_details)
-            scan_result[scan_name] = scan_details
-        except TimeoutError:
-            raise
-        except Exception as e:
-            # never fail but instead add an error message and keep an empty scan:
-            scan_result[scan_name] = []
-            messages = ['ERROR: ' + scan_name + ': ' + e.message]
-            if diag:
-                messages.append('ERROR: ' + scan_name + ': ' + traceback.format_exc())
-            scan_errors.extend(messages)
-
-    # put errors last, after scans proper
-    scan_result['scan_errors'] = scan_errors
-    return scan_result
-
-
-def save_results(scanners, files_count, results, format, options, input, output_file):
-    """
-    Save scan results to file or screen.
-    """
-
-    # note: in tests, sys.stdout is not used, but is instead some io
-    # wrapper with no name attributes. We use this to check if this is a
-    # real filesystem file or not.
-    # note: sys.stdout.name == '<stdout>' so it has a name.
-    is_real_file = hasattr(output_file, 'name')
-
-    if output_file != sys.stdout and is_real_file:
-        # we are writing to a real filesystem file: create directories!
-        parent_dir = os.path.dirname(output_file.name)
-        if parent_dir:
-            fileutils.create_dir(abspath(expanduser(parent_dir)))
-
-    # Write scan results to file or screen as a formatted output ...
-    # ... using a user-provided custom format template
-    format_plugins = plugincode.output.get_format_plugins()
-    if format not in format_plugins:
-        # format may be a custom template file path
-        if not os.path.isfile(format):
-            # this check was done before in the CLI validation, but this
-            # is done again if the function is used directly
-            echo_stderr('\nInvalid template: must be a file.', fg='red')
-        else:
-            from formattedcode import format_templated
-            # FIXME: carrying an echo function does not make sense
-            format_templated.write_custom(
-                results, output_file, _echo=echo_stderr, version=version, template_path=format)
+            error, value = interruptible(
+                partial(scanner.function, location), timeout=timeout)
+            if error:
+                msg = 'ERROR: for scanner: ' + scanner.key + ':\n' + error
+                errors.append(msg)
+            if value:
+                # a scanner function MUST return a sequence
+                scanner_result.extend(value)
+        except Exception:
+            msg = 'ERROR: for scanner: ' + scanner.key + ':\n' + traceback.format_exc()
+            errors.append(msg)
+        finally:
+            scan_time = time() - scan_time
```

Comment:
Shouldn't `scan_time` be calculated only once, outside of the `for` loop?

Reply:
This has been updated in the latest push. Thanks!